src/core/xmlHttpRequest.js

```diff
@@ -0,0 +1,112 @@
+ * @param {Object} options - fetch request options that implements RequestInit (https://fetch.spec.whatwg.org/#requestinit)
+ * @param {Function} [options.onUploadProgress]
+ * @param {Function} [options.onDownloadProgress]
+ * @returns {Promise<Response>}
+ */
+function xhr(url, options) {
+    return new Promise(resolve => {
+        const request = new XMLHttpRequest();
+        let responseBody = null;
+        const responseHeaders = new Headers();
+
+        if (typeof options.onUploadProgress === 'function') {
+            request.upload.addEventListener('progress', options.onUploadProgress);
+        }
+        if (typeof options.onDownloadProgress === 'function') {
+            request.addEventListener('progress', options.onDownloadProgress);
+        }
+        request.addEventListener('readystatechange', () => {
+            switch (request.readyState) {
+                case XHR_READY_STATE_OPENED:
+                    // eslint-disable-next-line no-case-declarations
+                    for (const header in options.headers) {
+                        request.setRequestHeader(header, options.headers[header]);
+                    }
+                    break;
+                case XHR_READY_STATE_HEADERS_RECEIVED:
+                    request.getAllResponseHeaders().trim().split(/[\r\n]+/).forEach((line) => {
+                        const parts = line.split(": ");
+                        const header = parts.shift();
+                        const value = parts.join(": ");
+                        if (header) {
+                            responseHeaders.append(header, value);
+                        }
+                    });
+                    break;
+                case XHR_READY_STATE_HEADERS_DONE:
+                    responseBody = request.response;
+                    // Response with null body status cannot have body
+                    if ([101, 204, 205, 304].includes(request.status)) {
+                        responseBody = null;
+                    }
+                    if (request.responseType === 'json') {
+                        responseBody = JSON.stringify(request.response);
+                    }
+                    // eslint-disable-next-line no-case-declarations
+                    const response = new Response(responseBody, {
+                        status: request.status,
+                        statusText: request.statusText,
+                        headers: responseHeaders
+                    });
+                    resolve(response);
+                    break;
+            }
+        });
+
+        request.open(options.method || 'GET', url, true);
+        request.send(options.body);
+    });
+}
+
+/**
+ * Creates an HTTP request to the url based on the provided parameters
+ * Request is based on fetch, so behaviour and parameters are the same, except
+ *   - every response where response code is not 2xx will be rejected and
+ *   - every response will be parsed as json.
```

Comment:
> Request is based on fetch

Is it the case here?

Reply:
Sort of, it implements the same options, return type, and `xhr` function will replicate the fetch behaviour, but using `XMLHttpRequest` under the hood